src/hooks/useOverlayManager.tsx

```diff
@@ -1,39 +1,67 @@
-import React, { useCallback, useContext, useState } from 'react'
+import React, { useCallback, useContext, useState, useRef } from 'react'
 import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock'
+import styled from '@emotion/styled'
 import { css, Global } from '@emotion/react'
+import { zIndex } from '@/shared/theme'
 
 type OverlayManagerContextValue = {
-  overlayOpen: boolean
-  setOverlayOpen: (value: boolean, scrollbarGap?: number) => void
+  scrollLocked: boolean
+  setScrollLocked: (value: boolean, scrollbarGap?: number) => void
+  overlayContainerOpened: boolean
+  setOverlayContainerOpened: (value: boolean) => void
+  overlayContainerRef: React.RefObject<HTMLDivElement>
 }
+
+type OverlayContainerProps = {
+  isOpened?: boolean
+}
+
 const OverlayManagerContext = React.createContext<OverlayManagerContextValue | undefined>(undefined)
 OverlayManagerContext.displayName = 'OverlayManagerContext'
 
 export const OverlayManagerProvider: React.FC = ({ children }) => {
-  const [overlayOpen, setOverlayOpen] = useState(false)
+  const [scrollLocked, setScrollLocked] = useState(false)
+  const [overlayContainerOpened, setOverlayContainerOpened] = useState(false)
   const [scrollbarGap, setScrollbarGap] = useState(0)
+  const overlayContainerRef = useRef<HTMLDivElement>(null)
 
-  const handleOverlayOpen = (value: boolean, scrollbarGap?: number) => {
-    if (value === overlayOpen) {
-      return
-    }
-
-    if (value) {
-      setOverlayOpen(true)
-      setScrollbarGap(scrollbarGap || 0)
-      disableBodyScroll(document.body, { reserveScrollBarGap: true })
-    } else {
-      setOverlayOpen(false)
-      setScrollbarGap(0)
-      enableBodyScroll(document.body)
-    }
-  }
+  const handleScrollLocked = useCallback(
+    (value: boolean, scrollbarGap?: number) => {
+      if (value === scrollLocked) {
+        return
+      }
+
+      if (value) {
+        setScrollLocked(true)
+        setScrollbarGap(scrollbarGap || 0)
+        disableBodyScroll(document.body, { reserveScrollBarGap: true })
+      } else {
+        setScrollLocked(false)
+        setScrollbarGap(0)
+        enableBodyScroll(document.body)
+      }
+    },
+    [scrollLocked]
```

Comment:
This still doesn't fix the issue I've described. Think about what happens with this as a dependency. Let's say we open the search overlay:
1) Overlay gets mounted, it contains this code:
```ts
  useEffect(() => {
    lockScroll()

    return () => {
      unlockScroll()
    }
  }, [lockScroll, unlockScroll])
```
So on mount the overlay will call `lockScroll()` from hook, that in turn will call `setScrollLocked` which is this function.
2) This function will change the value of `scrollLocked` among other things
3) Because `scrollLocked` is defined as a dependency here, `handleScrollLocked` will get updated with a new instance of the function
4) `lockScroll` in the hook has `handleScrollLocked` as a dependency so because of 3) it will get updated
5) the `useEffect` above has `lockScroll` from the hook as a dependency so because of 4) it will get updated

In summary, once the search overlay is mounted, we will call `lockScroll()`, `unlockScroll()` and then `lockScroll()` again or am I wrong?

We will need to remove that dependency here, I guess we can just drop that `if` at the beginning